packages/shell-api/src/database.ts

```diff
@@ -817,13 +817,36 @@ export default class Database extends ShellApiWithMongoClass {
       pipeline.push({ $match: matchingFilters });
     }
 
-    return (await this
-      .getSiblingDB('admin')
-      .aggregate(pipeline, {
-        '$readPreference': {
-          'mode': 'primaryPreferred'
-        }
-      })).toArray();
+    let result;
+
+    try {
+      const cursor = await this
+        .getSiblingDB('admin')
+        .aggregate(pipeline, {
+          $readPreference: {
+            mode: 'primaryPreferred'
+          }
+        });
+      result = await cursor.toArray();
+    } catch (error) {
+      if ((error as any).codeName === 'FailedToParse') {
+        const deepCopyPipeline = JSON.parse(JSON.stringify(pipeline));
+        delete deepCopyPipeline[0].$currentOp.truncateOps;
```

Comment:
Why are we cloning the pipeline here? I think that would make sense if the pipeline was passed in from the user, but we create it here inside this function, so it should be fine to just keep re-using it

(I’m bringing this up because the deep clone makes it seem like this could be any user-controlled input, but if it is user-controlled input, it could contain arbitrary BSON, and using JSON to clone that wouldn’t handle BSON values well, so this looks “wrong” to some degree without actually being wrong of course)

Reply:
I started from this, but it affects tests when truncateOps disappears not only for the second but also for the first call. And I think we want to track in tests that there are actually 2 calls that one rejects but the second resolves. What would be a better solution here?

Reply:
Maybe I also can assign $currentOp to a variable 🤔

Reply:
Yeah, thanks for the context. I think that’s something that the test should handle, though – you’re already modifying `serviceProvider.aggregateDb` with `.returns()`, you could also make that `.callsFake((...args) => ...)` and then store (a copy of) `args` each time? It’s partially re-implementing a bit of sinon functionality, yes, but I think that’s okay here